Ancestral sample

DL42 opened the pull request:
This adds new prototypes that combine mutation tracking and prior history and both with garbage collection (gc) (so multiple rounds of simplification with mutation and site tables).

Questions/Comments:

1) tested on full simulation with gc: more mutations generated from forward simulation than msprime over the same coalescent tree (possibly a bug in my script?) - seems to be more variance in msprime's number of mutations too (though harder to tell)

2) 
    """ 
    This is a mimic of a fwdpp
    recombination policy.

    We return a sorted list of breakpoints 
    on the interval [0,1).  The list is capped
    with the max value of a float (C/C++ double),
    which is a trick fwdpp uses.
    """"
    
    Should the above apply to mutations too? Why is this?
    
3) For mutations, how does fwdpp/fwdpy11 handle multiple mutation in same position? Should I search both the current tracker mutations and the mutations table each time I add a new mutation? 

4) I eliminated node_offset in population=nodes['population'] + node_offset  from both previous prototypes with prior_history

small oddity:

1) print(sites) sites table (and mutation table) prints nothing or 	for ancestral/derived state
rather than 0 when 0 is the value, prints 0 for a value of 48 (something to with int8?) which is what msprime seems to feed it by default - print(sites.ancestral_state) gives actual value

molpopgen commented:
Quick answers to questions first.

> """
> This is a mimic of a fwdpp
> recombination policy.
> 
> We return a sorted list of breakpoints
> on the interval [0,1). The list is capped
> with the max value of a float (C/C++ double),
> which is a trick fwdpp uses.
> """"
> 
> Should the above apply to mutations too? Why is this?

No need to do this for mutations.  For recombination, we copy stuff around up until the next point.  When the final xover is less than the last position in the offspring gamete, `numeric_limits<double>::max()` steps in, guaranteeing that all remaining mutation keys get copied.

> 
> For mutations, how does fwdpp/fwdpy11 handle multiple mutation in same position? Should I search both the current tracker mutations and the mutations table each time I add a new mutation?

It is not allows right now--strict infinitely-many sites w.r.to mutations currently in the simulation.  Internally, `std::unordered_set<double>` is used to handle this.

In your simulation, the easiest thing to do will be to use something like a `dict` to keep track of mutation positions.  Post-simplification, you can remove stuff from the dict no longer in the mutation table b/c it was simplified out.

DL42 commented:
> No need to do this for mutations. For recombination, we copy stuff around up until the next point. When the final xover is less than the last position in the offspring gamete, numeric_limits<double>::max() steps in, guaranteeing that all remaining mutation keys get copied.

Got it!

> It is not allows right now--strict infinitely-many sites w.r.to mutations currently in the simulation. Internally, std::unordered_set<double> is used to handle this.
> 
> In your simulation, the easiest thing to do will be to use something like a dict to keep track of mutation positions. Post-simplification, you can remove stuff from the dict no longer in the mutation table b/c it was simplified out.

Okay, that should work. Do you do this for the unordered_set currently? Remove mutations no longer in the simulation because the descendants of their generating node all died out?

molpopgen commented:
> Okay, that should work. Do you do this for the unordered_set currently? Remove mutations no longer in the simulation because the descendants of their generating node all died out?

Yes, although the solution is more brute-force, and linear in (no. gametes)*(no. mutations per gamete), which makes it one of fwdpp's least efficient [algorithms](https://github.com/molpopgen/fwdpp/blob/master/fwdpp/internal/sample_diploid_helpers.hpp#L15).

The results from the above link can be used to fill "recycling bins", which are indexes of fixed/extinct variants.  Those indexes are put in a FIFO queue, and new mutations/gametes go there first insead of getting push_back'd.

molpopgen commented on practice/prototype_mutations_history_gc.py:
```diff
@@ -0,0 +1,537 @@
+                                  child=tracker.edges['child'])
+        self.sites.append_columns(position=tracker.mutations['position'],
+                   ancestral_state=np.zeros(len(tracker.mutations['position']),np.int8),
+                   ancestral_state_length=np.ones(len(tracker.mutations['position']),np.uint32))
```
The `foo_length` is removed in the current msprime API.  The current master would have this line read as:

`ancestral_state_offset=np.arange(len(tracker.mutations['position'])+1,np.uint32)`

Similar for all occurrences of `derived_state_length`.

molpopgen commented on practice/prototype_mutations_history_gc.py:
```diff
@@ -0,0 +1,537 @@
+                                  parent=tracker.edges['parent'],
+                                  child=tracker.edges['child'])
+        self.sites.append_columns(position=tracker.mutations['position'],
+                   ancestral_state=np.zeros(len(tracker.mutations['position']),np.int8),
```
One thing to note (that I cannot figure out) is that you ancestral state is being printed out as -1 if you print the simplified mutations out to the screen.  However, the sites generated by msprime's mutation generator print out 0 as expected.

molpopgen replied:
(Actually, this may be a result of changes I made to get things working w/current msprime.  Thus, cannot rule out my introducing an error.)

DL42 replied:
When I update the newest version, I'll try to see if I can figure it out.

molpopgen commented:
OK, I have discovered something.

If you get rid of all periodic GC, then there's no meaningful difference in the length of the final two site tables.  Thus, there's an error somewhere in the simplifier.

DL42 commented:
>  OK, I have discovered something.
> 
> If you get rid of all periodic GC, then there's no meaningful difference in the length of the final two site tables. Thus, there's an error somewhere in the simplifier.

Yeah I was coming to a similar conclusion.

molpopgen commented:
> Yeah I was coming to a similar conclusion.

I have a branch, also called ancestral_sample, in my repo, where I've updated to the 0.5.0 API and introduced a mutation position lookup table in the form of a `dict`.

I also pinged Jerome via msprime and cc'd you--did you see that?  This is in regards to the `ancestral_state_offset` business.

I'm going to step aside for now--I cannot see what is going amiss, and perhaps a break will reset my brain.

DL42 commented:
> I have a branch, also called ancestral_sample, in my repo, where I've updated to the 0.5.0 API and introduced a mutation position lookup table in the form of a dict.
> 
> I also pinged Jerome via msprime and cc'd you--did you see that? This is in regards to the ancestral_state_offset business.
> 
> I'm going to step aside for now--I cannot see what is going amiss, and perhaps a break will reset my brain.

Yup, I see it.

jeromekelleher commented:
I'm happy to look into this; how do I reproduce this problem?

One thing that might be tripping you up is that ancestral and derived states are encoded as **ASCII characters**. So, if you want to have the **character** as the ancestral state, you need something like

```python
ancestral_state=np.zeros(len(tracker.mutations['position']),np.int8) + ord('0'),
```
ord('0') == 48, which is where this mysterious number comes from. I know it's not intuitive, but efficiently encoding variable length character data is a bloody pain, and this is the best I could do. This is the next thing to document (thoroughly!)...

molpopgen commented:
> 'm happy to look into this; how do I reproduce this problem?

If you clone this repo and checkout the ancestral_sample branch*, then run practice/prototype_mutations_history_gc.py with default params, with the current `msprime/master` installed and visible to your Python installation.

For e.g.,

```
PYTHONPATH=~/src/msprime python prototype_mutations_history_gc.py 
```

(wait a couple minutes....)

The output will be the the first 10 rows of our site table and then the first 10 from applying your mutation generator.

I tried the "ord" thing--didn't help.

I think there is something wrong with how we are updating the site/mutation tables while simplifying intermittently.  @DL42 noticed that we get more mutations than what your code outputs (`len(sites)` always `> len(sites2)`), but that if we set the GC interval to longer than the simulation length, that problem goes away. 

I'm also fairly certain I'm not updating the ancestral/derived state offsets properly during simplification during the sim...

DL42 commented:
> I think there is something wrong with how we are updating the site/mutation tables while simplifying intermittently. @DL42 noticed that we get more mutations than what your code outputs (len(sites) always > len(sites2)), but that if we set the GC interval to longer than the simulation length, that problem goes away.

Oddly I just ran the program with GC interval set to 10000 (it took a really long time) and it came up with more mutations, 106, in the forward simulation than msprime's 88 mutations - which is the same as it gave for the periodic gc. This is bizarre.

molpopgen commented:
> Oddly I just ran the program with GC interval set to 10000 (it took a really long time) and it came up with more mutations, 106, in the forward simulation than msprime's 88 mutations - which is the same as it gave for the periodic gc. This is bizarre.

Hmmm.  I'm doing that now.  My statement was based on setting the global `SIMLEN` to something like 5 and then setting the GC to be bigger.

molpopgen commented:
@DL42 -- yes, I see what you  are getting at, now.  I take it back.  We are goofing something in the recording of mutations.  Perhaps @jeromekelleher will spot it...

molpopgen commented:
I'm wondering about a potential ambiguity [here](http://msprime.readthedocs.io/en/latest/tutorial.html#working-with-tables), which states

```
 Each mutation is associated with a particular node (i.e., a particular ancestor), so that any sample which inherits from that node will also inherit that mutation, unless another mutation intervenes.
```

In terms of a forward sim, I interpret this to mean that the node ID of a mutation should be the gamete that first contained that mutation.  @jeromekelleher -- is this correct?

jeromekelleher commented:
> In terms of a forward sim, I interpret this to mean that the node ID of a mutation should be the gamete that first contained that mutation. @jeromekelleher -- is this correct?

This sounds right to me.

jeromekelleher commented:
I've looked though the code, and I don't see any obvious problems. I'm not sure I understand what the actual issue is though... The differences in the number of mutations seems within the range you'd expect to me. Is there a QQ plot or something showing a systematic difference?

jeromekelleher commented:
> A qqplot is here.

OK, yeah, that's way off!

> I still believe that I am indexing the ancestral/derived states incorrectly. For a site table of length n, and each state being 1 char, the state offsets should be np.arange(len(sites)+1, correct? That's not what we're doing...

That's correct yeah. What are you doing differently?

molpopgen commented:
> That's correct yeah. What are you doing differently?

[This](https://github.com/molpopgen/fwdpy11_arg_example/blob/ancestral_sample/practice/prototype_mutations_history_gc.py#L166).

I've tried various approaches to take the -1 value of the current data and then add to it.  But, I get either a "must be nrow+1" or "bad indexes" exception raised from msprime no matter what I try.

I think that the code linked to above is simply wrong, but it is passing through silently.

jeromekelleher commented:
The code looks fine to me. I've updated it to use the "ord('0')" trick, and it all seems to be doing what it should so (after printing out the tables and so on).

Have you tried loading up the two different tree sequences at the end and looking at where the mutations end up on the trees? (using tree.draw("file.svg")) 

I'd guess it's some subtle difference in how the mutations are being generated is the cause here.

molpopgen commented:
> The code looks fine to me. I've updated it to use the "ord('0')" trick, and it all seems to be doing what it should so (after printing out the tables and so on).

Where is your update?

molpopgen commented:
> The code looks fine to me. I've updated it to use the "ord('0')" trick, and it all seems to be doing what it should so (after printing out the tables and so on).

If I add `+ ord('0')`, as done [here](https://github.com/molpopgen/fwdpy11_arg_example/blob/ancestral_sample/practice/prototype_mutations_history_gc.py#L167), I still get output like this:

```
SiteTableRow(position=0.0021932576910003165, ancestral_state='\x00', metadata=b'')
SiteTableRow(position=0.02658839937361679, ancestral_state='\x00', metadata=b'')
SiteTableRow(position=0.026926033120680093, ancestral_state='\x00', metadata=b'')
SiteTableRow(position=0.04122585036401383, ancestral_state='\x00', metadata=b'')
SiteTableRow(position=0.04321784690986763, ancestral_state='\x00', metadata=b'')
SiteTableRow(position=0.06369870574085468, ancestral_state='\x00', metadata=b'')
SiteTableRow(position=0.09585633394422999, ancestral_state='\x00', metadata=b'')
SiteTableRow(position=0.12408964007903645, ancestral_state='\x00', metadata=b'')
SiteTableRow(position=0.1282390288787194, ancestral_state='\x00', metadata=b'')
SiteTableRow(position=0.1529373725621489, ancestral_state='\x00', metadata=b'')
```

DL42 commented:
> Have you tried loading up the two different tree sequences at the end and looking at where the mutations end up on the trees? (using tree.draw("file.svg"))

> I'd guess it's some subtle difference in how the mutations are being generated is the cause here.

This gives me an idea ... the root node will contain all the mutations that have "fixed" in the population/sample. These are not segregating in the sample, but were generated along the tree. They aren't eliminated in this version of gc, because they still exist (we're only eliminating mutations than have gone to 0). So msprime's simplification step will put them on the root of the tree. When msprime simulates mutations, there should be none at the root node as after the simulation is done, there is no time before the root to place mutations on.

I'll check this to see if this is the discrepancy.

molpopgen commented:
> This gives me an idea ... the root node will contain all the mutations that have "fixed" in the population/sample.

This should be testable via metadata, where you track the origin time of each mutation and ask how many pre-date the root.

molpopgen commented:
> Have you tried loading up the two different tree sequences at the end and looking at where the mutations end up on the trees? (using tree.draw("file.svg"))

I'm not able to get mutation info to show up.  mutation_labels and mutation_locations are not affecting the resulting svg files.

molpopgen commented:
> Have you tried loading up the two different tree sequences at the end and looking at where the mutations end up on the trees? (using tree.draw("file.svg"))
> 
> I'm not able to get mutation info to show up. mutation_labels and mutation_locations are not affecting the resulting svg files.

@jeromekelleher -- ignore that!  I was looking at a plot for a marginal tree w/no mutations.  Still digging in...

molpopgen commented:
@jeromekelleher Is it possible that attempting to draw a tree with mutations on the root will trigger an exception?  I'm getting indexing errors when attempting to `draw` some of the marginal trees.

DL42 commented:
> >This gives me an idea ... the root node will contain all the mutations that have "fixed" in the population/sample.

> This should be testable via metadata, where you track the origin time of each mutation and ask how many pre-date the root.

So what I did was to simply compare the number of mutations assigned to the root node. For the default parameters there were indeed 26 mutations at the root after simplifying the tables from the forward simulation and 0 at the root from msprime generating mutations from the tree. The difference in total mutations is 106 to 88. The difference in non-root mutations is thus 80 to 88. 

I'll push the commits so you can take a look.

molpopgen commented:
> I'll push the commits so you can take a look. I think we have the explanation for the excess mutations. Would it be possible to do another QQ plot?

I'll take a look ASAP.  Definitely will get back to you by Monday.

jeromekelleher commented:
> Where is your update?

[here](https://github.com/jeromekelleher/fwdpy11_arg_example/blob/ord-trick/practice/prototype_mutations_history_gc.py#L170)

jeromekelleher commented:
> @jeromekelleher Is it possible that attempting to draw a tree with mutations on the root will trigger an exception? I'm getting indexing errors when attempting to draw some of the marginal trees.

Entirely possible --- the tree drawing stuff isn't very battle-hardened.

jeromekelleher commented:
> o what I did was to simply compare the number of mutations assigned to the root node for the tree corresponding to the genomic interval the mutation fell into. These mutations had to occur at or before the root node's existence.

Sounds like a plausible explanation to me.

DL42 commented:
> Where is your update?
here

I changed our code in the latest commit to match yours more closely and the printing of the ancestral/derived states works now ... it seems that the key is that `len(tracker.mutations['position']) + ord('0'), np.int8)` (what we had) produces a different result from `len(tracker.mutations['position']), np.int8) + ord('0')`.

DL42 commented:
Nope, I think the changes I currently have planned are best suited for a future merge.

molpopgen commented:
OK, this is merged into dev.  master is currently where we are w.r.to the paper.  This PR is the only difference right now b/w dev and master.  @DL42, you can base subsequent PRs from dev.

DL42 commented:
Got it!

molpopgen commented:
Thanks for the help @jeromekelleher!

jeromekelleher commented:
No worries, happy to help!
